src/Streamly/Internal/Data/Array/Foreign.hs

```diff
@@ -195,18 +196,41 @@ fromStream (SerialT m) = P.fold A.write m
 -- Elimination
 -------------------------------------------------------------------------------
 
+{-# INLINE_NORMAL producer #-}
+producer :: forall m a. (Monad m, Storable a) => Producer m (Array a) a
+producer = Producer step inject extract
+    where
+
+    inject (Array (ForeignPtr start contents) (Ptr end)) =
+        return $ MA.ReadUState (ForeignPtr end contents) (Ptr start)
+
+    {-# INLINE_LATE step #-}
+    step (MA.ReadUState fp@(ForeignPtr end _) p)
+        | assert (p <= Ptr end) (p == Ptr end) =
+            let x = unsafeInlineIO $ touchForeignPtr fp
+            in x `seq` return D.Stop
+    step (MA.ReadUState fp p) = do
+            -- unsafeInlineIO allows us to run this in Identity monad for pure
+            -- toList/foldr case which makes them much faster due to not
+            -- accumulating the list and fusing better with the pure consumers.
+            --
+            -- This should be safe as the array contents are guaranteed to be
+            -- evaluated/written to before we peek at them.
+            let !x = unsafeInlineIO $ peek p
+            return $ D.Yield x
+                (MA.ReadUState fp (p `plusPtr` sizeOf (undefined :: a)))
+
+    extract (MA.ReadUState (ForeignPtr end contents) (Ptr p)) =
+        return $ Array (ForeignPtr p contents) (Ptr end)
+
 -- | Unfold an array into a stream.
 --
 -- /Since 0.7.0 (Streamly.Memory.Array)/
 --
 -- @since 0.8.0
 {-# INLINE_NORMAL read #-}
 read :: forall m a. (Monad m, Storable a) => Unfold m (Array a) a
-read = Unfold.lmap A.unsafeThaw MA.read
-
-{-# INLINE_NORMAL producer #-}
-producer :: forall m a. (Monad m, Storable a) => Producer m (Array a) a
-producer = Producer.translate A.unsafeThaw A.unsafeFreeze MA.producer
+read = Producer.simplify producer
```

Comment:
similar code exists in "src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs"

Reply:
In src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs we no longer use unsafeInlineIO. Till I figure out its implication on the performance, I've just duplicated it here.